### ISSUE TITLE
[DX] Add a utility-like function that removes contrib modules that were merged into core.

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#3760]
This is a follow-up to #3742, and related to #3653

@klonos 
> ...wondering if it's worth moving the code we are reusing into a `backdrop_remove_merged_modules()` function, and then calling that from the update hooks (we are going to be merging more modules as time goes by - and repetition of the very exact code hurts my eyes 😅).

@quicksketch 
> That's a great idea @klonos! We have a set of utility-like functions in `includes/update.inc` that includes things like `update_variable_get()` and `update_module_enable()`. Creating an `update_module_uninstall()` or even `update_module_uninstall_merged_module()` could get rid of redundancy.

---
PR by @klonos: https://github.com/backdrop/backdrop/pull/2647

**Comment by klonos** [comment]
PR up for review. Suggesting 1.12.7 as the milestone.

**Comment by klonos** [comment]
Here's a thing that we have not considered: what happens if the main contrib module we are uninstalling has submodules, which are also enabled?...

I have been testing the messages in my code by calling the newly-added `update_uninstall_merged_module()` function, and passing `'devel'` as the module name. I was doing this with the line that executed the `db_query("DELETE FROM...` bit commended out (so that there is no actual module uninstallation while testing). Before committing and pushing my changes, I have removed the `//` from the commented out line of course, but I accidentally left the call to `update_uninstall_merged_module('devel')` change to another file (which was not part of the commit). I saw the warnings about removing the Devel module, and that it is not in `backdrop_merged_modules()`, and I had a face-palming moment. ...to my surprise though, the devel module was not uninstalled. ...head-scratching moment ...then I realised that I also had `devel_generate` enabled, which requires `devel`.

So what I think has happened here is that because there was a requirement on `devel` by an enabled module, the `devel` module got enabled again.

Am I thinking this the wrong way, or should we be accounting for that scenario as well? What would the best approach be in that case? Should we be iterating through the module dependencies, and building a list of modules that need to be uninstalled/disabled along with the "main" contrib module? ...or am I overthinking this perhaps?

**Comment by opi** [comment]
I'm OK with a "non-plural" function to avoid over complicated code.

IF (how many times?) we have to deal with multiple modules, we'll foreach.
IF we have to deal with submodules, we'll know that, and have to write the corresponding code in the relevant hook_update_N.

kiss

**Comment by klonos** [comment]
>IF we have to deal with submodules, we'll know that, and have to write the corresponding code in the relevant hook_update_N.

👍 to that ...with the exception that IF/when this need comes to be, we should make changes to the `update_uninstall_merged_module()` function, so it can be reused.

Another thing that I like about having a function for this now, is that we now have a single place for the messages. So this is standardised + translators have only a single string to translate.

**Comment by opi** [comment]
> +1 to that ...with the exception that IF/when this need comes to be, we should make changes to the update_uninstall_merged_module() function, so it can be reused.

Nope, the `foreach` and ` if ($submodules)` parts will be handled in specifique `hook_update_N`.

With your example of devel+devel_generate. Imagine that Backdrop include a devel-like in core, then the  `system_update_1013` will take care of uninstalling devel_generate first and then uninstall devel.

Your nice `update_uninstall_merged_module()` function still be the basic "uninstall ONE merged module" without any business logic.

**Comment by klonos** [comment]
...I have updated the inline comments in the PR to better explain why we are using direct db queries.

The new `update_uninstall_merged_module()` function returns `TRUE`/`FALSE` depending on whether the module was found or not. I want to retain this, but what if I wanted it to also return a message (a string - not call `backdrop_set_message()`), what would be the best approach to do that? ...I was thinking something like a second parameter, passed by reference. Like this:

`function update_uninstall_merged_module($module, &$message)`

Any pointers?

**Comment by jenlampton** [comment]
> Should we be iterating through the module dependencies, and building a list of modules that need to be uninstalled/disabled along with the "main" contrib module? ...or am I overthinking this perhaps?

Don't we already have something in core that does something similar? Drush is able to uninstall all dependencies when a parent module is uninstalled. Is it using core functions to do that, or does it have its own?

> ...I have updated the inline comments in the PR to better explain why we are using direct db queries.

Unless you are doing a query that has (or may have) access control a `db_query()` is the way to go since it's so much faster. Code comment probably not necessary :)

> but what if I wanted it to also return a message

Why not test the true/false in the caller and add the message there instead? Why would you prefer that it come from the uninstall action?
